### PR TITLE
Update task model

### PR DESCRIPTION
**Pull request by schote**
Required for https://github.com/brain-link/scanhub-ui/pull/34 

Update of task model, artifacts, arguments and directories are now of type `dict[string, string]`

### CODE REVIEW

**Review by BenWilhelmUniklinikFreiburg — COMMENTED** [review]
I tried the merge locally:
- did the suggested merge
- checked out the main branch in scanhub-ui
- removed parcel cache
- removed db docker volume
- rebuilt all docker images
- run yarn install --check-files
- started the software

Then I get the following error in the frontend:
grafik

I guess it is supposed to be used with a newer frontend version. I'll try that next.

**Comment by BenWilhelmUniklinikFreiburg** [comment]
Similar error with the latest scanhub-ui commit 3e55800 from branch scanhub-ui/issue/33-login-functionality

**Comment by schote** [comment]
> I tried the merge locally:
> 
> * did the suggested merge
> * checked out the main branch in scanhub-ui
> * removed parcel cache
> * removed db docker volume
> * rebuilt all docker images
> * run yarn install --check-files
> * started the software
> 
> Then I get the following error in the frontend: grafik
> 
> I guess it is supposed to be used with a newer frontend version. I'll try that next.

Thank you for testing, did you regenerate the open api client for the latest scanhub commit? For any change in the pydantic model, the client needs to be re-generated. This could explain the type error, as some types of the task model have changed as described above. Also don't forget to rebuild the base image (and the exam-manager with the new base).

**Comment by schote** [comment]
I tested this branch with the latest scanhub-ui branch from the [open PR](https://github.com/brain-link/scanhub-ui/pull/34) with the following steps:
1. Delete base-image, exam-manager, user-manager, and scanhub-db (storing users and tasks)
2. Build the base image
3. Compose scanhub
4. With all containers running regenerate the clients (yarn generate-all-clients)
5. Delete parcel-cache, dist, and node-modules
6. Run yarn install and yarn start

Bildschirmfoto 2024-06-07 um 23 01 45

File: services/base/shared_libs/src/scanhub_libraries/models.py [hunk 1]
```diff
@@ -231,8 +232,10 @@ class Config:
     description: str
     type: TaskType
     args: dict[str, str]
-    artifacts: dict[str, list[dict[str, str]]]
-    task_destinations: list[dict[str, str]]
+    artifacts: dict[str, str]
+    destinations: dict[str, str]
+    # artifacts: dict[str, list[dict[str, str]]]
+    # task_destinations: list[dict[str, str]]
```
Comment (jobehrens) [inline services/base/shared_libs/src/scanhub_libraries/models.py:238]: maybe better to remove when unused?

File: services/exam-manager/app/db.py [hunk 2]
```diff
@@ -128,10 +128,14 @@ class Task(Base):
     # Input and output artifacts with export destination
     # Example: {"input": [{"name": "env_HOLOSCAN_INPUT_PATH", "value": "{{ context.input.dicom }}"}]}
     # https://github.com/Project-MONAI/monai-deploy/blob/main/deploy/monai-deploy-express/sample-workflows/hello-world.json
-    artifacts: Mapped[dict[str, list[dict[str, str]]]] = mapped_column(type_=JSON, nullable=True) # implemention of input and output artifacts for artifact types as replacement for string
-
-    # List of task destinations, which are for example used to create the chain of topics in the kafka message broker, i.e., target topics
-    task_destinations: Mapped[list[dict[str, str]]] = mapped_column(type_=JSON, nullable=True)
+    # implemention of input and output artifacts for artifact types as replacement for string
+    artifacts: Mapped[dict[str, str]] = mapped_column(type_=JSON, nullable=True)
+    # artifacts: Mapped[dict[str, list[dict[str, str]]]] = mapped_column(type_=JSON, nullable=True)
```
Comment (jobehrens) [inline services/exam-manager/app/db.py:133]: remove?

File: services/exam-manager/app/db.py [hunk 3]
```diff
@@ -128,10 +128,14 @@ class Task(Base):
...
+
+    # List of task destinations, which are for example used to create the chain of topics in the kafka message broker,
+    # i.e., target topics
+    destinations: Mapped[dict[str, str]] = mapped_column(type_=JSON, nullable=True)
+    # task_destinations: Mapped[list[dict[str, str]]] = mapped_column(type_=JSON, nullable=True)
```
Comment (jobehrens) [inline services/exam-manager/app/db.py:138]: remove?

File: services/base/shared_libs/src/scanhub_libraries/models.py (hunk 1 above)
Comment (schote) [reply services/base/shared_libs/src/scanhub_libraries/models.py:238]: You are right, we have it in the git history

File: services/exam-manager/app/db.py (hunk 3 above)
Comment (schote) [reply services/exam-manager/app/db.py:138]: I can update it later today, but feel free to delete the commented lines 👍🏼

**Comment by BenWilhelmUniklinikFreiburg** [comment]
Ok, the rebuild for the merged main branch of the backend, that is the result of this merge request, works now for me too. However the frontend main is far behind the backend main.

Regenerating the clients should in principle not be necessary, because the clients are committed to git. But I see why it was necessary here.

When updating the outdated frontend main by merging scanhub-ui/issue/33-login-functionality and rebuilding, I get some regular error with the latest feature of scanhub-ui/issue/33-login-functionality. Updating the backend main further by merging scanhub-ui/issue/33-login-functionality again and removing the scanhub_postgres_data_scanhub volume to reset the database columns I get an up to date and working version.

**Review by jobehrens — APPROVED** [review]

**Pull request merged**